activation/post_test.go

```diff
@@ -49,31 +49,47 @@ func TestPostSetupManager(t *testing.T) {
 	})
 
 	// Create data.
-	req.NoError(mgr.StartSession(context.Background(), mgr.opts))
+	cfg, err := mgr.InitializerConfig(context.Background(), mgr.opts)
+	req.NoError(err)
+	req.NoError(mgr.StartSession(context.Background(), cfg))
 	cancel()
 	_ = eg.Wait()
 
 	req.Equal(PostSetupStateComplete, mgr.Status().State)
 
 	// Create data (same opts).
-	req.NoError(mgr.StartSession(context.Background(), mgr.opts))
+	req.NoError(mgr.StartSession(context.Background(), cfg))
 
 	// Cleanup.
 	req.NoError(mgr.Reset())
 
 	// Create data (same opts, after deletion).
-	req.NoError(mgr.StartSession(context.Background(), mgr.opts))
+	req.NoError(mgr.StartSession(context.Background(), cfg))
 	req.Equal(PostSetupStateComplete, mgr.Status().State)
 }
 
-func TestPostSetupManager_StateError(t *testing.T) {
+// Checks that InitializerConfig returns an error when invalid opts are given.
+// It's not exhaustive since this validation occurs in the post repo codebase
+// and should be fully tested there.
+func TestPostSetupManager_InitializerConfig(t *testing.T) {
 	req := require.New(t)
 
 	mgr := newTestPostManager(t)
-	mgr.opts.NumUnits = 0
-	req.Error(mgr.StartSession(context.Background(), mgr.opts))
-	// Verify Status returns StateError
-	req.Equal(PostSetupStateError, mgr.Status().State)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	// check no error with good options.
+	_, err := mgr.InitializerConfig(ctx, mgr.opts)
+	req.NoError(err)
+
+	cfg := config.DefaultConfig()
+
+	// Check that invalid options return errors
+	opts := mgr.opts
+	opts.NumUnits = cfg.MaxNumUnits + 1
+	_, err = mgr.InitializerConfig(ctx, opts)
+	req.Error(err)
```

Comment:
Could `InitializerConfig` be renamed to `ValidatePostOpts` or similar and just return an error if the provided Options are invalid?

I feel like the reason it returns a new object that holds additional data and a new instance of the `Initializer` exposes too much of the internals of the `PostSetupManager`.

I think the logic could be simplified if `commitmentAtx` would just be a part of `PostSetupOpts` that 1) if set gets checked to be a valid ATX and 2) if not set defaults to the logic how it is selected at the moment

Reply:
> This change sounds like it's not related to not failing loudly, so my inclination would be to not include it in this PR.

I agree that this is out of scope for this PR.

> ... there would be now multiple places in the code where a commitmentAtx is chosen, and optional config fields tend to complicate systems because you now have multiple different ways of calling/configuring components.

The `PostSetupOpts` contain client side configuration of the PoST the node uses. This includes the nodes public key, the number of space units to use, the directory the data is stored in aso. Since a node is free to choose any ATX as their commitment it should imo be part of these options too.

The selected commitment ATX is even stored along the other options in the `post_metadata.json` file and if the operator chooses to initialize using our Post CLI has to be set explicitly already.

Just because we have code that selects a commitment ATX if the operator doesn't explicitly choose one, should not cause it to be handled differently from other configuration. This behaviour is the same for other parameters too: eg. the post directory can be set explicitly by the operator, but if not the node will select the user directory (based on the OS the node is running on).

Reply:
> I'm not sure what you mean by exposing the internals, the struct is exported but the fields of the struct are not?

I think this would make sense: call the `InitializerConfig` method `NewSession`, the returned objects type `PostSession` and (if possible) drop the `commitmentAtx` and `PostSetupOpts` from it (they are already part of the `Initializer` and as far as I can tell are only needed to log them when `StartSession` is called.

Reply:
The naming sounds sensible to me. But the `PostSetupOpts` are needed because `PostSetupManger.lastOpts` are set to them and the `commitmentAtxId` is needed because `PostSetupManager.commitmentAtxId` is set to it. If we update those fields in `IntializerConfig/NewSession` then we need to go for a more complicated locking model to ensure that they can't be modified while `StartSession` is running.

Reply:
The way I think about the `InitializationConfig/PostSession` is that its a black box, its just a means of neatly ensuring sequential calls to first configure and then run intialization. It's not really important what is in it because its all just an implementation detail of the `PostSetupManager` and in fact we can make `InitializationConfig/PostSession` unexported. I only didn't because the the methods of PostSetupManager are all exported, but now I realise that is only so that they can be mocked.

Reply:
> the PostSetupOpts are needed because `PostSetupManger.lastOpts` are set to them and the `commitmentAtxId` is needed because `PostSetupManager.commitmentAtxId` is set to it

That's true and I guess we have to leave it like this for now, because changing it is out of scope of this PR. My thoughts on it just for future reference:

- `commitmentAtxId` besides logging is only used in `GenerateProof` where it is only passed in to check if the post metadata contains the same `commitmentAtxId`. It is not needed to generate a proof - only for initialization and the same check already happens when we call `StartSession` (indirectly via `NewInitializer`).
- `lastOps` is also only needed in 2 places - to get the size of the PoST during proof generation in `activation/activation.go` and as part of the return value of `Status()`
  - the first one has to be changed - the `GenerateProof` method should return the number of SU it used to generate the proof. That number might be less than the value used during initialization, because we lost data since then (due to disk failures), or it might be more (if we want to allow PoST size increases while the node is running).
  - The second one needs to be thought through again - the data in `lastOpts` is mostly only needed during initialization and is not really valid afterwards. Is there even a need to return network parameters like `ScryptParams` or initialization parameters (like the device used to initialize the data, the size of the batches used during initialization and if we throttle during initialization) as part of the PostStatus? If yes should those still be returned after initialization finished? At the moment we don't even return the values used during initialization, but the ones `StartSession` were called with, which might be different if the node was restarted at some point.

Reply:
Added ticket to fix in future here: https://github.com/spacemeshos/go-spacemesh/issues/4336

Reply:
Thanks